frontends/p4/sideEffects.cpp

```diff
@@ -391,39 +432,282 @@ const IR::Node* DoSimplifyExpressions::preorder(IR::MethodCallExpression* mce) {
         }
     }
 
-    if (anyOut) {
-        // Check aliasing between all pairs of arguments where at
-        // least one of them is out or inout.
-        for (auto p1 : *mi->substitution.getParametersInArgumentOrder()) {
-            auto arg1 = mi->substitution.lookup(p1);
-            if (hasSideEffects.count(arg1->expression))
+    // Check aliasing between all pairs of arguments where at
+    // least one of them is out or inout.
+    for (auto p1 : *mi->substitution.getParametersInArgumentOrder()) {
+        if (p1->direction == IR::Direction::None)
+           continue;
+        auto arg1 = mi->substitution.lookup(p1);
+
+        for (auto p2 : *mi->substitution.getParametersInArgumentOrder()) {
+            if (p2->direction == IR::Direction::None)
                 continue;
-            for (auto p2 : *mi->substitution.getParametersInArgumentOrder()) {
-                LOG3("p1=" << dbp(p1) << " p2=" << dbp(p2));
-                if (p2 == p1)
-                    break;
-                if (!p1->hasOut() && !p2->hasOut())
+
+            LOG3("p1=" << dbp(p1) << " p2=" << dbp(p2));
+            if (p2 == p1)
+                break;
+
+            auto arg2 = mi->substitution.lookup(p2);
+
+            if (auto m = arg1->expression->to<IR::Member>()) {
+                auto u = arg1->expression->to<IR::Operation_Unary>();
+                if  (u->expr->is<IR::ArrayIndex>()) {
+                    auto b = u->expr->to<IR::Operation_Binary>();
+                    // we handle this case T f(in T x, inout T y){...}; T g(inout T z){...}
+                    // f(g(val), s[val].x)
+                    // tmp_0 = val;
+                    // tmp_1 = s[tmp_0].x;
+                    // tmp_2 = g(val);
+                    // f(tmp_2, tmp_1);
+                    // s[tmp_0].x = tmp_1;
+                    if (!b->right->is<IR::MethodCallExpression>() &&
+                        arg2->expression->is<IR::MethodCallExpression>()) {
+                        auto mce2 = arg2->expression->to<IR::MethodCallExpression>();
+                        auto p3 = findAlias(p1, b->right, mce2);
+                        if  (p3) {
+                            auto mi2 = MethodInstance::resolve(mce2, refMap, typeMap);
+                            auto arg3 = mi2->substitution.lookup(p3);
```

Comment:
I don't understand why arg3 is not null. You seem to assume that p3 is the same as p2 here. What if the call you are analyzing has more than 2 arguments and the aliasing one is the third one?

Reply:
I didn't understand the question.
Maybe the name (arg3) is wrong.
arg3 is an argument of g what an alias with val. 
It is not third argument of f.
Maybe you didn't mean that.

here are 2 examples:

extern bit f(in bit x, inout bit y);
extern bit g(inout bit x, inout bit y);

f(g(b, a), s[a].z);

after SideEffectOrdering:
tmp = a_0;
tmp_0 = s_0[tmp].z;
tmp_1 = g(b_0, a_0);
f(tmp_1, tmp_0);
s_0[tmp].z = tmp_0;

extern bit f(inout bit x, in bit y, in bit z);
extern bit g(inout bit x);

f(s[a].z, b, g(a));
after SideEffectOrdering:
tmp = a_0;
tmp_0 = s_0[tmp].z;
tmp_1 = g(a_0);
f(tmp_0, b_0, tmp_1);
s_0[tmp].z = tmp_0;